src/p11_child/p11_child_openssl.c

```diff
@@ -0,0 +1,1133 @@
+/*
+    SSSD
+
+    Helper child to commmunicate with SmartCard via OpenSSL
+
+    Authors:
+        Sumit Bose <sbose@redhat.com>
+
+    Copyright (C) 2018 Red Hat
+
+    This program is free software; you can redistribute it and/or modify
+    it under the terms of the GNU General Public License as published by
+    the Free Software Foundation; either version 3 of the License, or
+    (at your option) any later version.
+
+    This program is distributed in the hope that it will be useful,
+    but WITHOUT ANY WARRANTY; without even the implied warranty of
+    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+    GNU General Public License for more details.
+
+    You should have received a copy of the GNU General Public License
+    along with this program.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+#include <errno.h>
+#include <openssl/ssl.h>
+#include <openssl/crypto.h>
+#include <openssl/x509.h>
+#include <openssl/err.h>
+#include <openssl/rand.h>
+#include <p11-kit/p11-kit.h>
+
+#include <popt.h>
+
+#include "util/util.h"
+#include "util/crypto/sss_crypto.h"
+#include "util/child_common.h"
+#include "p11_child/p11_child.h"
+
+struct p11_ctx {
+    X509_STORE *x509_store;
+    const char *ca_db;
+};
+
+static int talloc_cleanup_openssl(struct p11_ctx *p11_ctx)
+{
+    CRYPTO_cleanup_all_ex_data();
+    //OPENSSL_INIT_free(openssl_init_settings);
+
+    return 0;
+}
+errno_t init_p11_ctx(TALLOC_CTX *mem_ctx, const char *ca_db,
+                     struct p11_ctx **p11_ctx)
+{
+    struct p11_ctx *ctx;
+
+    //OPENSSL_INIT_SETTINGS *openssl_init_settings = NULL;
+
+    //openssl_init_settings = OPENSSL_init_new();
+    SSL_library_init();
```

Comment:
Here it would be nice to add a comment that this call expands to openssl_init_ssl with the new openssl versions, because when I looked at the man page of SSL_library_init, I noticed that this function is deprecated and noticed the expansion only when I was looking at the openssl headers to find out why there is no deprecation warning..